pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	// TODO: support sending IGMP query with both v2 and v3 messages
+	queryVersion      = 3
+	podInterfaceIndex = "podInterface"
+
+	// How long to wait before retrying the processing of a multicast group change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+)
+
+var (
+	workerCount uint8 = 2
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers is a set for the local Pod's interface name which has joined in the multicast group.
+	localMembers sets.String
+	lastProbe    time.Time
```

Comment:
Should it be called "lastIGMPReport"? It is not the "probe" (IGMP query) time right?

Reply:
updated.